Dockerfile

```diff
@@ -33,6 +22,16 @@ RUN npm install --location=global yarn
 
 WORKDIR /opt/visual-qontract
 
+### test image
+FROM base AS test
+
+ENV CI=1
+
+RUN yarn install && yarn run lint && yarn test 
+
+### prod image
+FROM base AS prod
+
```

Comment:
I'm thinking to simplify the whole file with 3 stages, no node runtime in the final image, and bump to use `ubi9` like this

```dockerfile
FROM registry.access.redhat.com/rhel9/nodejs-20 AS base
# is root only used to install yarn?
USER root
ENV CI=1
RUN npm install --location=global yarn
WORKDIR /opt/visual-qontract
COPY . .
RUN yarn install
RUN yarn build

FROM base as test
RUN yarn lint && yarn test

FROM registry.access.redhat.com/ubi9/nginx-124 AS prod
# copy nginx conf, entrypoint and build results
# may need to use COPY --chown for matching permissions
COPY deployment/entrypoint.sh /
COPY deployment/nginx.conf.template /etc/nginx/nginx.conf.template
COPY --from=base /opt/visual-qontract/build/ /opt/visual-qontract/build/
EXPOSE 8080
USER 1001
ENTRYPOINT ["/entrypoint.sh"]
```

not tested, need to verify permissions

Reply:
Thanks!

Reply:
I updated the Dockerfile following your suggestions and it worked, however I ran into the `ERR_OSSL_EVP_UNSUPPORTED` error when trying to switch from ubi8 to ubi9. Created a separate ticket here for that: https://issues.redhat.com/browse/APPSRE-11046

Reply:
Additionally, I built the new image w/ my local Docker setup on my PC against commercial app-interface and saw no errors.

Reply:
`ERR_OSSL_EVP_UNSUPPORTED` is because ubi9 upgraded openssl to 3 while ubi8 is using openssl 1.1, we need to either upgrade webpack version (part of `react-scripts`), or set `NODE_OPTIONS=--openssl-legacy-provider` env to build.

To install `yarn` using `npm`, should it be `npm install --global yarn` according to [doc](https://classic.yarnpkg.com/lang/en/docs/install/), or there are some special need in ubi?

Another best practice is to use `--frozen-lockfile` to ensure no update to lockfile.

Reply:
btw usually `devDependencies` are required during `yarn build`, we can just do it without `--production` and skip another install in `test`

Reply:
Updated to address your suggestions, ty! I was going down the route of updating `react-scripts` but then quickly ran into other dependency issues :upside_down_face: 

However, adding the `--openssl-legacy-provider` option worked

Reply:
The correct syntax is `npm install --global yarn`, if omit `--`, it means to install this package https://www.npmjs.com/package/global and yarn locally. Btw seems root permission is not needed to install yarn. Maybe the default user is good enough for everything.

If you want to speed up build time via cache, can split package install and build into 2 steps

```dockerfile
COPY package.json yarn.lock ./
RUN yarn install --non-interactive --frozen-lockfile
COPY . .
RUN yarn build
```

This way any changes outside of package.json or lock can reuse build cache.

Here https://github.com/app-sre/visual-qontract/pull/226/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R17, no need to run `yarn` again, can just do `yarn lint && yarn test` to avoid reinstall.

And why `--openssl-legacy-provider build` added to every scripts, should it just need `--openssl-legacy-provider` part?

Reply:
Whoops :facepalm: on the npm install